File: lib/services/mykad_sdk/model/my_kad_model.dart
class MyKadModel {
  String? name;
  String? gmpcName;
  String? kptName;
  String? icNo;
  String? oldIcNo;
  String? dob;
  String? pob;
  String? gender;
  String? citizenship;
  String? issueDate;
  String? race;
  String? religion;
  String? address1;
  String? address2;
  String? address3;
  String? postCode;
  String? city;
  String? state;
  String? photo;

  MyKadModel({
    this.name,
    this.gmpcName,
    this.kptName,
    this.icNo,
    this.oldIcNo,
    this.dob,
    this.pob,
    this.gender,
    this.citizenship,
    this.issueDate,
    this.race,
    this.religion,
    this.address1,
    this.address2,
    this.address3,
    this.postCode,
    this.city,
    this.state,
    this.photo,
  });

  // fromJson method
  factory MyKadModel.fromJson(Map<String, dynamic> json) {
    Map<String, dynamic> mykadData = json['mykad'];
    return MyKadModel(
      name: mykadData['name'],
      gmpcName: mykadData['gmpc_name'],
      kptName: mykadData['kpt_name'],
      icNo: mykadData['icno'],
      oldIcNo: mykadData['old_icno'],
      dob: mykadData['date_of_birth'],
      pob: mykadData['place_of_birth'],
      gender: mykadData['gender'],
      citizenship: mykadData['citizenship'],
      issueDate: mykadData['issue_date'],
      race: mykadData['race'],
      religion: mykadData['religion'],
      address1: mykadData['address1'],
      address2: mykadData['address2'],
      address3: mykadData['address3'],
      postCode: mykadData['poscode'],
      city: mykadData['city'],
      state: mykadData['state'],
      photo: mykadData['photo'],
    );
  }

  // toJson method
  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'gmpc_name': gmpcName,
      'kpt_name': kptName,
      'icno': icNo,
      'old_icno': oldIcNo,
      'date_of_birth': dob,
      'place_of_birth': pob,
      'gender': gender,
      'citizenship': citizenship,
      'issue_date': issueDate,
      'race': race,
      'religion': religion,
      'address1': address1,
      'address2': address2,
      'address3': address3,
      'poscode': postCode,
      'city': city,
      'state': state,
      'photo': photo,
    };
  }

  // copyWith method
  MyKadModel copyWith({
    String? name,
    String? gmpcName,
    String? kptName,
    String? icNo,
    String? oldIcNo,
    String? dob,
    String? pob,
    String? gender,
    String? citizenship,
    String? issueDate,
    String? race,
    String? religion,
    String? address1,
    String? address2,
    String? address3,
    String? postCode,
    String? city,
    String? state,
    String? photo,
  }) {
    return MyKadModel(
      name: name ?? this.name,
      gmpcName: gmpcName ?? this.gmpcName,
      kptName: kptName ?? this.kptName,
      icNo: icNo ?? this.icNo,
      oldIcNo: oldIcNo ?? this.oldIcNo,
      dob: dob ?? this.dob,
      pob: pob ?? this.pob,
      gender: gender ?? this.gender,
      citizenship: citizenship ?? this.citizenship,
      issueDate: issueDate ?? this.issueDate,
      race: race ?? this.race,
      religion: religion ?? this.religion,
      address1: address1 ?? this.address1,
      address2: address2 ?? this.address2,
      address3: address3 ?? this.address3,
      postCode: postCode ?? this.postCode,
      city: city ?? this.city,
      state: state ?? this.state,
      photo: photo ?? this.photo,
    );
  }
}

File: android/app/src/main/java/com/example/mykad_sdk/MyKadModel.java
package com.example.mykad_sdk;

import android.util.Log;

import com.securemetric.reader.myid.MYKAD;

import org.json.JSONObject;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

public class MyKadModel {
    private String name;
    private String gmpcName;
    private String kptName;
    private String icNo;
    private String oldIcNo;
    private Date dob;
    private String pob;
    private String gender;
    private String citizenship;
    private Date issueDate;
    private String race;
    private String religion;
    private String address1;
    private String address2;
    private String address3;
    private String postCode;
    private String city;
    private String state;
    private String photo;

    public static String mykadToJson(MYKAD myKAD, String photo) {
        JSONObject jMyKad = new JSONObject();
        JSONObject jMyKadDetail = new JSONObject();
        SimpleDateFormat sdt = new SimpleDateFormat("yyyy-MM-dd z", Locale.getDefault());

        try {
            jMyKadDetail.put("name", myKAD.getName());
            jMyKadDetail.put("gmpc_name", myKAD.getGmpcName());
            jMyKadDetail.put("kpt_name", myKAD.getKptName());
            jMyKadDetail.put("icno", myKAD.getIcNo());
            jMyKadDetail.put("old_icno", myKAD.getOldIcNo());

            Date dob = myKAD.getDob();
            if (dob != null) {
                try {
                    sdt.setTimeZone(TimeZone.getTimeZone("GMT+8"));
                    jMyKadDetail.put("date_of_birth", sdt.format(dob));
                } catch (Exception e) {
                    Log.w("MyKAD_SDK", Log.getStackTraceString(e));

                }
            }

            jMyKadDetail.put("place_of_birth", myKAD.getPob());
            jMyKadDetail.put("gender", myKAD.getGender());
            jMyKadDetail.put("citizenship", myKAD.getCitizenship());

            Date issueDate = myKAD.getIssueDate();
            if (issueDate != null) {
                try {
                    sdt.setTimeZone(TimeZone.getTimeZone("GMT+8"));
                    jMyKadDetail.put("issue_date", sdt.format(issueDate));
                } catch (Exception e) {
                    Log.w("MyKAD_SDK", Log.getStackTraceString(e));

                }
            }

            jMyKadDetail.put("race", myKAD.getRace());
            jMyKadDetail.put("religion", myKAD.getReligion());
            jMyKadDetail.put("address1", myKAD.getAddress1());
            jMyKadDetail.put("address2", myKAD.getAddress2());
            jMyKadDetail.put("address3", myKAD.getAddress3());
            jMyKadDetail.put("poscode", myKAD.getPostCode());
            jMyKadDetail.put("city", myKAD.getCity());
            jMyKadDetail.put("state", myKAD.getState());

            if (photo != null) {
                photo = photo.replace("\n", "");
                jMyKadDetail.put("photo", photo);
            }

            jMyKad.put("mykad", jMyKadDetail);

            return jMyKad.toString();
        } catch (Exception e) {
            Log.e("MyKAD_SDK", Log.getStackTraceString(e));
            return "";
        }
    }
}
